### PR TITLE
Feature/dubbo32. resteasy filter&intercept support

### PR DESCRIPTION
**Pull request by suncairong163**
## What is the purpose of the change
add resteasy filter&intercept support

## Checklist
- [x] Make sure there is a [GitHub_issue](https://github.com/apache/dubbo/issues) field for the change (usually before you start working on it). Trivial changes like typos do not require a GitHub issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
- [ ] Each commit in the pull request should have a meaningful subject line and body.
- [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [ ] Check if is necessary to patch to Dubbo 3 if you are work on Dubbo 2.7
- [ ] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add sample in [dubbo samples](https://github.com/apache/dubbo-samples) project.
- [ ] Add some description to [dubbo-website](https://github.com/apache/dubbo-website) project if you are requesting to add a feature.
- [ ] GitHub Actions works fine on your own branch.
- [ ] If this contribution is large, please follow the [Software Donation Guide](https://github.com/apache/dubbo/wiki/Software-donation-guide).

### CODE REVIEW

**Comment by suncairong163** [comment]
@AlbumenJ  @CrazyHZM

**Comment by chickenlj** [comment]
Please add the newly added SPI extension to shade configuration in dubbo-all/pom.xml and fix the conflicts.

```
2023-06-11T15:10:59.4344411Z [ERROR] Failures: 
2023-06-11T15:10:59.4346957Z [ERROR]   FileTest.checkDubboAllTransform:379 Newly created SPI interface must be added to dubbo-all(dubbo-distribution/dubbo-all/pom.xml in shade plugin) to being transformed. Found spis: [org.apache.dubbo.rpc.protocol.rest.filter.RestFilter, org.apache.dubbo.rpc.protocol.rest.filter.RestResponseInterceptor] ==> expected: <true> but was: <false>
```

File: dubbo-rpc/dubbo-rpc-rest/src/main/java/org/apache/dubbo/rpc/protocol/rest/filter/ServiceInvokeRestResponseInterceptor.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+@Activate(value = "invoke",order = Integer.MAX_VALUE)
+public class ServiceInvokeRestResponseInterceptor implements RestResponseInterceptor {
+
+    @Override
+    public void intercept(URL url, RequestFacade request, NettyHttpResponse nettyHttpResponse, Object result, RpcInvocation rpcInvocation, RestResponseInterceptorChain interceptorChain, ServiceDeployer serviceDeployer) throws Exception {
+
+        writeResult(nettyHttpResponse, request, url, result, rpcInvocation.getReturnType());
+        nettyHttpResponse.setStatus(200);
+    }
+}
```
Comment (CrazyHZM) [inline dubbo-rpc/dubbo-rpc-rest/src/main/java/org/apache/dubbo/rpc/protocol/rest/filter/ServiceInvokeRestResponseInterceptor.java:41]: This does not need to exist and may skew other non-200 results.

File: dubbo-rpc/dubbo-rpc-rest/src/main/java/org/apache/dubbo/rpc/protocol/rest/filter/ServiceInvokeRestFilter.java [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+            }
+        } else {
+
+            // invoke the intercept chain before Result  write to  response
+            new RestResponseInterceptor() {
+                @Override
+                public void intercept(URL url, RequestFacade request, NettyHttpResponse response, Object result, RpcInvocation rpcInvocation, RestResponseInterceptorChain interceptorChain, ServiceDeployer serviceDeployer) throws Exception {
+                    interceptorChain.intercept(url, request, response, result, rpcInvocation, interceptorChain, serviceDeployer);
+                }
+            }.intercept(url, request, nettyHttpResponse, result.getValue(), rpcInvocation, new RestResponseInterceptorChain(restResponseInterceptors), serviceDeployer);
```
Comment (CrazyHZM) [inline dubbo-rpc/dubbo-rpc-rest/src/main/java/org/apache/dubbo/rpc/protocol/rest/filter/ServiceInvokeRestFilter.java:126]: 1. Just RestResponseInterceptor cycle call, I think it is not necessary to increase RestResponseInterceptorChain.
2. RestResponseInterceptor should also handle exceptions, so you should not only support RestResponseInterceptor when there are no exceptions

File: dubbo-rpc/dubbo-rpc-rest/src/main/java/org/apache/dubbo/rpc/protocol/rest/extension/resteay/filter/ResteasyContainerFilterAdapter.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+import org.apache.dubbo.rpc.protocol.rest.request.RequestFacade;
+import org.jboss.resteasy.specimpl.BuiltResponse;
+
+import javax.ws.rs.container.ContainerRequestFilter;
+import java.util.List;
+
+import static org.apache.dubbo.common.constants.CommonConstants.RESTEASY_NETTY_HTTP_REQUEST_ATTRIBUTE_KEY;
+
+
+@Activate(value = "resteasy", onClass = {"javax.ws.rs.container.ContainerRequestFilter", "org.jboss.resteasy.plugins.server.netty.NettyHttpRequest"}, order = Integer.MAX_VALUE - 1)
```
Comment (CrazyHZM) [inline dubbo-rpc/dubbo-rpc-rest/src/main/java/org/apache/dubbo/rpc/protocol/rest/extension/resteay/filter/ResteasyContainerFilterAdapter.java:36]: There is also a lack of compatibility with `javax.ws.rs.container.ContainerResponseFilter`

**Review by CrazyHZM — CHANGES_REQUESTED** [review]

**Comment by suncairong163** [comment]
@CrazyHZM  change request & response filter

File: dubbo-rpc/dubbo-rpc-rest/src/main/java/org/apache/dubbo/rpc/protocol/rest/extension/resteasy/intercept/DubboServerWriterInterceptorContext.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+public class DubboServerWriterInterceptorContext extends ServerWriterInterceptorContext {
+    public DubboServerWriterInterceptorContext(WriterInterceptor[] interceptors, ResteasyProviderFactory providerFactory, Object entity, Class type, Type genericType, Annotation[] annotations, MediaType mediaType, MultivaluedMap<String, Object> headers, OutputStream outputStream, HttpRequest request) {
+        super(interceptors, providerFactory, entity, type, genericType, annotations, mediaType, headers, outputStream, request);
+    }
+
+
+    @Override
+    public void proceed() throws IOException, WebApplicationException {
+        LogMessages.LOGGER.debugf("Interceptor Context: %s,  Method : proceed", getClass().getName());
```
Comment (CrazyHZM) [inline dubbo-rpc/dubbo-rpc-rest/src/main/java/org/apache/dubbo/rpc/protocol/rest/extension/resteasy/intercept/DubboServerWriterInterceptorContext.java:41]: Replace with dubbo's own logger.

File: dubbo-rpc/dubbo-rpc-rest/src/main/java/org/apache/dubbo/rpc/protocol/rest/filter/RestFilterManager.java [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+import org.apache.dubbo.rpc.protocol.rest.deploy.ServiceDeployer;
+import org.apache.dubbo.rpc.protocol.rest.filter.context.RestFilterContext;
+import org.apache.dubbo.rpc.protocol.rest.filter.context.RestInterceptContext;
+import org.apache.dubbo.rpc.protocol.rest.netty.NettyHttpResponse;
+import org.apache.dubbo.rpc.protocol.rest.request.RequestFacade;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class RestFilterManager {
```
Comment (CrazyHZM) [inline dubbo-rpc/dubbo-rpc-rest/src/main/java/org/apache/dubbo/rpc/protocol/rest/filter/RestFilterManager.java:31]: Each method should go back to where it was originally used and static classes are not a very good approach.
For example, executeRequestFilters and executeResponseFilters should revert to NettyHttpHandler, and restRequestFilters should be fetched from the FrameworkModel in the context. Such as the url FrameworkModel.

**Review by CrazyHZM — CHANGES_REQUESTED** [review]

**Comment by suncairong163** [comment]
@CrazyHZM  fix some  change request

File: dubbo-rpc/dubbo-rpc-rest/src/main/java/org/apache/dubbo/rpc/protocol/rest/extension/resteasy/intercept/DubboServerWriterInterceptorContext.java (hunk 4 above)
Comment (CrazyHZM) [reply dubbo-rpc/dubbo-rpc-rest/src/main/java/org/apache/dubbo/rpc/protocol/rest/extension/resteasy/intercept/DubboServerWriterInterceptorContext.java:41]: ping @suncairong163

**Review by CrazyHZM — APPROVED** [review]

**Pull request merged**